test/solver/solver.cpp

```diff
@@ -0,0 +1,742 @@
+        gko::test::make_lower_triangular(data);
+    }
+
+    static typename solver_type::parameters_type build(
+        std::shared_ptr<const gko::Executor> exec,
+        gko::size_type iteration_count)
+    {
+        return solver_type::build();
+    }
+
+    static typename solver_type::parameters_type build_preconditioned(
+        std::shared_ptr<const gko::Executor>, gko::size_type)
+    {
+        assert(false);
+    }
+};
+
+
+struct UpperTrs : SimpleSolverTest<gko::solver::UpperTrs<solver_value_type>> {
+    static bool is_iterative() { return false; }
+
+    static bool is_preconditionable() { return false; }
+
+    static double tolerance() { return r<value_type>::value; }
+
+    static void preprocess(gko::matrix_data<value_type, index_type>& data)
+    {
+        // make sure the diagonal is nonzero
+        gko::test::make_hpd(data, 1.2);
+        gko::test::make_upper_triangular(data);
+    }
+
+    static typename solver_type::parameters_type build(
+        std::shared_ptr<const gko::Executor> exec,
+        gko::size_type iteration_count)
+    {
+        return solver_type::build();
+    }
+
+    static typename solver_type::parameters_type build_preconditioned(
+        std::shared_ptr<const gko::Executor>, gko::size_type)
+    {
+        assert(false);
+    }
+};
+
+
+template <typename ObjectType>
+struct test_pair {
+    std::shared_ptr<ObjectType> ref;
+    std::shared_ptr<ObjectType> dev;
+
+    test_pair(std::unique_ptr<ObjectType> ref_obj,
+              std::shared_ptr<const gko::Executor> exec)
+        : ref{std::move(ref_obj)}, dev{gko::clone(exec, ref)}
+    {}
+
+    test_pair(std::unique_ptr<ObjectType> ref_obj,
+              std::unique_ptr<ObjectType> dev_obj)
+        : ref{std::move(ref_obj)}, dev{std::move(dev_obj)}
+    {}
+};
+
+
+template <typename T>
+class Solver : public ::testing::Test {
+protected:
+    using Config = T;
+    using SolverType = typename T::solver_type;
+    using Mtx = typename T::matrix_type;
+    using index_type = gko::int32;
+    using value_type = typename Mtx::value_type;
+    using mixed_value_type = gko::next_precision<value_type>;
+    using Vec = gko::matrix::Dense<value_type>;
+    using MixedVec = gko::matrix::Dense<mixed_value_type>;
+
+    Solver() : rand_engine(15) {}
+
+    void SetUp()
+    {
+        ref = gko::ReferenceExecutor::create();
+        init_executor(ref, exec);
+    }
+
+    void TearDown()
+    {
+        if (exec != nullptr) {
+            ASSERT_NO_THROW(exec->synchronize());
+        }
+    }
+
+    test_pair<Mtx> gen_mtx(int num_rows, int num_cols, int min_cols,
+                           int max_cols)
+    {
+        auto data =
+            gko::test::generate_random_matrix_data<value_type, index_type>(
+                num_rows, num_cols,
+                std::uniform_int_distribution<>(min_cols, max_cols),
+                std::normal_distribution<>(0.0, 1.0), rand_engine);
+        Config::preprocess(data);
```

Comment:
You could just use `generate_random_matrix<Mtx>`, once the matrix_utils are templated on the matrix type. The functions below don't actually need the intermediate step of generating a `matrix_data` outside the `generate_random_matrix` function.

Reply:
see https://github.com/ginkgo-project/ginkgo/pull/973#discussion_r810464108